From 570290c7416cfae57e4d59b703afec0a018a9ae6 Mon Sep 17 00:00:00 2001
From: Pradyot Bhattacharyya <pbhattacharyya@habana.ai>
Date: Sat, 10 May 2025 09:28:42 +0300
Subject: [PATCH] [SW-226833] gaudi3: correct i2c configuration

Configure i2c configuration to the following:
i2c0: 400Khz
i2c1: 270Khz
i2c2: 100Khz
This change is required because by default,
i2c clocks are misconfigured to run at 270Khz.
We are keeping i2c1 at 270 for consistency.
Also we are using the FAST configuration
for configuring 270KHz because FAST is for
speeds <=400 KHz

Change-Id: Iee9726f8151b624c26f5dec1fee4e76d7bfc7dfc
Signed-off-by: Pradyot Bhattacharyya <pbhattacharyya@habana.ai>
---

diff --git a/boards/hl/gaudi3/hl_gaudi3_preboot_asic.dts b/boards/hl/gaudi3/hl_gaudi3_preboot_asic.dts
index f9c3f62..0c379ec 100644
--- a/boards/hl/gaudi3/hl_gaudi3_preboot_asic.dts
+++ b/boards/hl/gaudi3/hl_gaudi3_preboot_asic.dts
@@ -141,7 +141,6 @@
 };
 
 &i2c2 {
-	clock-frequency = <I2C_BITRATE_FAST_PLUS>;
 	status = "okay";
 };
 
diff --git a/drivers/i2c/CMakeLists.txt b/drivers/i2c/CMakeLists.txt
index ca761ab..276a4c7 100644
--- a/drivers/i2c/CMakeLists.txt
+++ b/drivers/i2c/CMakeLists.txt
@@ -105,3 +105,4 @@
 
 add_subdirectory_ifdef(CONFIG_I2C_TARGET target)
 add_subdirectory_ifdef(CONFIG_HL_I2C_SHIM_DRIVER	hl_i2c)
+zephyr_include_directories(${DEVICE})
\ No newline at end of file
diff --git a/drivers/i2c/fs1/i2c_dw.h b/drivers/i2c/fs1/i2c_dw.h
new file mode 100644
index 0000000..b109dda
--- /dev/null
+++ b/drivers/i2c/fs1/i2c_dw.h
@@ -0,0 +1,223 @@
+/* dw_i2c.h - header for Design Ware I2C operations */
+
+/*
+ * Copyright (c) 2015 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_DRIVERS_I2C_I2C_DW_H_
+#define ZEPHYR_DRIVERS_I2C_I2C_DW_H_
+
+#include <zephyr/drivers/i2c.h>
+#include <stdbool.h>
+
+#define DT_DRV_COMPAT snps_designware_i2c
+
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(pcie)
+BUILD_ASSERT(IS_ENABLED(CONFIG_PCIE), "DW I2C in DT needs CONFIG_PCIE");
+#include <zephyr/drivers/pcie/pcie.h>
+#endif
+
+#if defined(CONFIG_RESET)
+#include <zephyr/drivers/reset.h>
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define I2C_DW_MAGIC_KEY			0x44570140
+
+
+typedef void (*i2c_isr_cb_t)(const struct device *port);
+
+
+#define IC_ACTIVITY                     (1 << 0)
+#define IC_ENABLE_BIT                   (1 << 0)
+
+
+/* dev->state values from IC_DATA_CMD Data transfer mode settings (bit 8) */
+#define I2C_DW_STATE_READY                 (0)
+#define I2C_DW_CMD_SEND                    (1 << 0)
+#define I2C_DW_CMD_RECV                    (1 << 1)
+#define I2C_DW_CMD_ERROR                   (1 << 2)
+#define I2C_DW_BUSY                        (1 << 3)
+
+
+#define DW_ENABLE_TX_INT_I2C_MASTER		(DW_INTR_STAT_TX_OVER |  \
+						 DW_INTR_STAT_TX_EMPTY | \
+						 DW_INTR_STAT_TX_ABRT |  \
+						 DW_INTR_STAT_STOP_DET)
+#define DW_ENABLE_RX_INT_I2C_MASTER		(DW_INTR_STAT_RX_UNDER | \
+						 DW_INTR_STAT_RX_OVER |  \
+						 DW_INTR_STAT_RX_FULL | \
+						 DW_INTR_STAT_STOP_DET)
+
+#define DW_ENABLE_TX_INT_I2C_SLAVE		(DW_INTR_STAT_RD_REQ | \
+						 DW_INTR_STAT_TX_ABRT | \
+						 DW_INTR_STAT_STOP_DET)
+#define DW_ENABLE_RX_INT_I2C_SLAVE		(DW_INTR_STAT_RX_UNDER | \
+						 DW_INTR_STAT_RX_FULL | \
+						 DW_INTR_STAT_RX_DONE | \
+						 DW_INTR_STAT_STOP_DET)
+
+#define DW_DISABLE_ALL_I2C_INT		0x00000000
+
+
+/* IC_CON Low count and high count default values */
+/* TODO verify values for high and fast speed */
+#define I2C_STD_HCNT			(CONFIG_I2C_DW_CLOCK_SPEED * 4)
+#define I2C_STD_LCNT			(CONFIG_I2C_DW_CLOCK_SPEED * 5)
+#define I2C_FS_HCNT			((CONFIG_I2C_DW_CLOCK_SPEED * 6) / 8)
+#define I2C_FS_LCNT			((CONFIG_I2C_DW_CLOCK_SPEED * 7) / 8)
+#define I2C_HS_HCNT			((CONFIG_I2C_DW_CLOCK_SPEED * 6) / 8)
+#define I2C_HS_LCNT			((CONFIG_I2C_DW_CLOCK_SPEED * 7) / 8)
+#define I2C_CUSTOM_HCNT			I2C_FS_HCNT
+#define I2C_CUSTOM_LCNT			I2C_FS_LCNT
+
+/*
+ * DesignWare speed values don't directly translate from the Zephyr speed
+ * selections in include/i2c.h so here we do a little translation
+ */
+#define I2C_DW_SPEED_STANDARD		0x1
+#define I2C_DW_SPEED_FAST		0x2
+#define I2C_DW_SPEED_FAST_PLUS		0x2
+#define I2C_DW_SPEED_HIGH		0x3
+
+/*
+ * SDA Holdtime conversion from nanosecond to clock ticks.
+ * (0.5*10^6) added before division with 10^6 to avoid
+ * losing fraction.
+ */
+#define i2c_dw_ns_to_clks(x) (((x * CONFIG_I2C_DW_CLOCK_SPEED * 1000) + 500000)\
+								/ 1000000)
+
+
+/*
+ * These values have been randomly selected.  It would be good to test different
+ * watermark levels for performance capabilities
+ */
+#define I2C_DW_TX_WATERMARK		2
+#define I2C_DW_RX_WATERMARK		7
+
+
+struct i2c_dw_rom_config {
+	DEVICE_MMIO_ROM;
+	i2c_isr_cb_t	config_func;
+	uint32_t		bitrate;
+
+#if defined(CONFIG_PINCTRL)
+	const struct pinctrl_dev_config *pcfg;
+#endif
+#if defined(CONFIG_RESET)
+	const struct reset_dt_spec reset;
+#endif
+
+	uint32_t		sda_hold_time;
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(pcie)
+	struct pcie_dev *pcie;
+#endif /* I2C_DW_PCIE_ENABLED */
+
+#ifdef CONFIG_I2C_DW_LPSS_DMA
+	const struct device *dma_dev;
+#endif
+};
+
+struct i2c_dw_dev_config {
+	DEVICE_MMIO_RAM;
+	struct k_sem		device_sync_sem;
+	struct k_mutex		bus_mutex;
+	uint32_t app_config;
+
+	uint8_t			*xfr_buf;
+	uint32_t		xfr_len;
+	uint32_t		rx_pending;
+	bool			user_abort_pending;
+
+	uint16_t		hcnt;
+	uint16_t		lcnt;
+
+	volatile uint8_t	state;  /* last direction of transfer */
+	uint32_t		request_bytes;
+	uint8_t			xfr_flags;
+	bool			support_hs_mode;
+#ifdef CONFIG_I2C_DW_LPSS_DMA
+	uintptr_t phy_addr;
+	uintptr_t base_addr;
+	/* For dma transfer */
+	bool xfr_status;
+#endif
+
+	uint32_t		sda_hold_time_val;
+#ifdef CONFIG_I2C_TARGET
+	struct i2c_target_config *slave_cfg;
+	bool slave_attached;
+#endif
+
+	struct {
+		uint32_t	enable;
+		uint32_t	con;
+		uint32_t	intr_stat;
+		uint32_t	raw_intr_stat;
+		uint32_t	status;
+		uint32_t	tx_abrt_src;
+		uint32_t	rxtl;
+		uint32_t	txtl;
+		uint32_t	rxflr;
+		uint32_t	txflr;
+		uint32_t	comp_param_1;
+	} regs;
+	const char *failure_reason;
+};
+
+#define Z_REG_READ(__sz) sys_read##__sz
+#define Z_REG_WRITE(__sz) sys_write##__sz
+#define Z_REG_SET_BIT sys_set_bit
+#define Z_REG_CLEAR_BIT sys_clear_bit
+#define Z_REG_TEST_BIT sys_test_bit
+
+#define DEFINE_MM_REG_READ(__reg, __off, __sz)				\
+	static inline uint32_t read_##__reg(uintptr_t addr)			\
+	{								\
+		return Z_REG_READ(__sz)(addr + __off);			\
+	}
+#define DEFINE_MM_REG_WRITE(__reg, __off, __sz)				\
+	static inline void write_##__reg(uint32_t data, uintptr_t addr)	\
+	{								\
+		Z_REG_WRITE(__sz)(data, addr + __off);			\
+	}
+
+#define DEFINE_SET_BIT_OP(__reg_bit, __reg_off, __bit)			\
+	static inline void set_bit_##__reg_bit(uintptr_t addr)		\
+	{								\
+		Z_REG_SET_BIT(addr + __reg_off, __bit);			\
+	}
+
+#define DEFINE_CLEAR_BIT_OP(__reg_bit, __reg_off, __bit)		\
+	static inline void clear_bit_##__reg_bit(uintptr_t addr)		\
+	{								\
+		Z_REG_CLEAR_BIT(addr + __reg_off, __bit);		\
+	}
+
+#define DEFINE_TEST_BIT_OP(__reg_bit, __reg_off, __bit)			\
+	static inline int test_bit_##__reg_bit(uintptr_t addr)		\
+	{								\
+		return Z_REG_TEST_BIT(addr + __reg_off, __bit);		\
+	}
+
+#ifdef __cplusplus
+}
+#endif
+
+#if defined(CONFIG_I2C_DW_DEV_RECOVERY)
+int i2c_check_and_recover(const struct device *dev);
+void i2c_platform_spec_init(const struct device *dev);
+#else
+static inline int i2c_check_and_recover(const struct device *dev)
+{
+	return 1;
+}
+static void i2c_platform_spec_init(const struct device *dev) {}
+#endif
+
+#endif /* ZEPHYR_DRIVERS_I2C_I2C_DW_H_ */
diff --git a/drivers/i2c/gaudi2/i2c_dw.h b/drivers/i2c/gaudi2/i2c_dw.h
new file mode 100644
index 0000000..b109dda
--- /dev/null
+++ b/drivers/i2c/gaudi2/i2c_dw.h
@@ -0,0 +1,223 @@
+/* dw_i2c.h - header for Design Ware I2C operations */
+
+/*
+ * Copyright (c) 2015 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_DRIVERS_I2C_I2C_DW_H_
+#define ZEPHYR_DRIVERS_I2C_I2C_DW_H_
+
+#include <zephyr/drivers/i2c.h>
+#include <stdbool.h>
+
+#define DT_DRV_COMPAT snps_designware_i2c
+
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(pcie)
+BUILD_ASSERT(IS_ENABLED(CONFIG_PCIE), "DW I2C in DT needs CONFIG_PCIE");
+#include <zephyr/drivers/pcie/pcie.h>
+#endif
+
+#if defined(CONFIG_RESET)
+#include <zephyr/drivers/reset.h>
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define I2C_DW_MAGIC_KEY			0x44570140
+
+
+typedef void (*i2c_isr_cb_t)(const struct device *port);
+
+
+#define IC_ACTIVITY                     (1 << 0)
+#define IC_ENABLE_BIT                   (1 << 0)
+
+
+/* dev->state values from IC_DATA_CMD Data transfer mode settings (bit 8) */
+#define I2C_DW_STATE_READY                 (0)
+#define I2C_DW_CMD_SEND                    (1 << 0)
+#define I2C_DW_CMD_RECV                    (1 << 1)
+#define I2C_DW_CMD_ERROR                   (1 << 2)
+#define I2C_DW_BUSY                        (1 << 3)
+
+
+#define DW_ENABLE_TX_INT_I2C_MASTER		(DW_INTR_STAT_TX_OVER |  \
+						 DW_INTR_STAT_TX_EMPTY | \
+						 DW_INTR_STAT_TX_ABRT |  \
+						 DW_INTR_STAT_STOP_DET)
+#define DW_ENABLE_RX_INT_I2C_MASTER		(DW_INTR_STAT_RX_UNDER | \
+						 DW_INTR_STAT_RX_OVER |  \
+						 DW_INTR_STAT_RX_FULL | \
+						 DW_INTR_STAT_STOP_DET)
+
+#define DW_ENABLE_TX_INT_I2C_SLAVE		(DW_INTR_STAT_RD_REQ | \
+						 DW_INTR_STAT_TX_ABRT | \
+						 DW_INTR_STAT_STOP_DET)
+#define DW_ENABLE_RX_INT_I2C_SLAVE		(DW_INTR_STAT_RX_UNDER | \
+						 DW_INTR_STAT_RX_FULL | \
+						 DW_INTR_STAT_RX_DONE | \
+						 DW_INTR_STAT_STOP_DET)
+
+#define DW_DISABLE_ALL_I2C_INT		0x00000000
+
+
+/* IC_CON Low count and high count default values */
+/* TODO verify values for high and fast speed */
+#define I2C_STD_HCNT			(CONFIG_I2C_DW_CLOCK_SPEED * 4)
+#define I2C_STD_LCNT			(CONFIG_I2C_DW_CLOCK_SPEED * 5)
+#define I2C_FS_HCNT			((CONFIG_I2C_DW_CLOCK_SPEED * 6) / 8)
+#define I2C_FS_LCNT			((CONFIG_I2C_DW_CLOCK_SPEED * 7) / 8)
+#define I2C_HS_HCNT			((CONFIG_I2C_DW_CLOCK_SPEED * 6) / 8)
+#define I2C_HS_LCNT			((CONFIG_I2C_DW_CLOCK_SPEED * 7) / 8)
+#define I2C_CUSTOM_HCNT			I2C_FS_HCNT
+#define I2C_CUSTOM_LCNT			I2C_FS_LCNT
+
+/*
+ * DesignWare speed values don't directly translate from the Zephyr speed
+ * selections in include/i2c.h so here we do a little translation
+ */
+#define I2C_DW_SPEED_STANDARD		0x1
+#define I2C_DW_SPEED_FAST		0x2
+#define I2C_DW_SPEED_FAST_PLUS		0x2
+#define I2C_DW_SPEED_HIGH		0x3
+
+/*
+ * SDA Holdtime conversion from nanosecond to clock ticks.
+ * (0.5*10^6) added before division with 10^6 to avoid
+ * losing fraction.
+ */
+#define i2c_dw_ns_to_clks(x) (((x * CONFIG_I2C_DW_CLOCK_SPEED * 1000) + 500000)\
+								/ 1000000)
+
+
+/*
+ * These values have been randomly selected.  It would be good to test different
+ * watermark levels for performance capabilities
+ */
+#define I2C_DW_TX_WATERMARK		2
+#define I2C_DW_RX_WATERMARK		7
+
+
+struct i2c_dw_rom_config {
+	DEVICE_MMIO_ROM;
+	i2c_isr_cb_t	config_func;
+	uint32_t		bitrate;
+
+#if defined(CONFIG_PINCTRL)
+	const struct pinctrl_dev_config *pcfg;
+#endif
+#if defined(CONFIG_RESET)
+	const struct reset_dt_spec reset;
+#endif
+
+	uint32_t		sda_hold_time;
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(pcie)
+	struct pcie_dev *pcie;
+#endif /* I2C_DW_PCIE_ENABLED */
+
+#ifdef CONFIG_I2C_DW_LPSS_DMA
+	const struct device *dma_dev;
+#endif
+};
+
+struct i2c_dw_dev_config {
+	DEVICE_MMIO_RAM;
+	struct k_sem		device_sync_sem;
+	struct k_mutex		bus_mutex;
+	uint32_t app_config;
+
+	uint8_t			*xfr_buf;
+	uint32_t		xfr_len;
+	uint32_t		rx_pending;
+	bool			user_abort_pending;
+
+	uint16_t		hcnt;
+	uint16_t		lcnt;
+
+	volatile uint8_t	state;  /* last direction of transfer */
+	uint32_t		request_bytes;
+	uint8_t			xfr_flags;
+	bool			support_hs_mode;
+#ifdef CONFIG_I2C_DW_LPSS_DMA
+	uintptr_t phy_addr;
+	uintptr_t base_addr;
+	/* For dma transfer */
+	bool xfr_status;
+#endif
+
+	uint32_t		sda_hold_time_val;
+#ifdef CONFIG_I2C_TARGET
+	struct i2c_target_config *slave_cfg;
+	bool slave_attached;
+#endif
+
+	struct {
+		uint32_t	enable;
+		uint32_t	con;
+		uint32_t	intr_stat;
+		uint32_t	raw_intr_stat;
+		uint32_t	status;
+		uint32_t	tx_abrt_src;
+		uint32_t	rxtl;
+		uint32_t	txtl;
+		uint32_t	rxflr;
+		uint32_t	txflr;
+		uint32_t	comp_param_1;
+	} regs;
+	const char *failure_reason;
+};
+
+#define Z_REG_READ(__sz) sys_read##__sz
+#define Z_REG_WRITE(__sz) sys_write##__sz
+#define Z_REG_SET_BIT sys_set_bit
+#define Z_REG_CLEAR_BIT sys_clear_bit
+#define Z_REG_TEST_BIT sys_test_bit
+
+#define DEFINE_MM_REG_READ(__reg, __off, __sz)				\
+	static inline uint32_t read_##__reg(uintptr_t addr)			\
+	{								\
+		return Z_REG_READ(__sz)(addr + __off);			\
+	}
+#define DEFINE_MM_REG_WRITE(__reg, __off, __sz)				\
+	static inline void write_##__reg(uint32_t data, uintptr_t addr)	\
+	{								\
+		Z_REG_WRITE(__sz)(data, addr + __off);			\
+	}
+
+#define DEFINE_SET_BIT_OP(__reg_bit, __reg_off, __bit)			\
+	static inline void set_bit_##__reg_bit(uintptr_t addr)		\
+	{								\
+		Z_REG_SET_BIT(addr + __reg_off, __bit);			\
+	}
+
+#define DEFINE_CLEAR_BIT_OP(__reg_bit, __reg_off, __bit)		\
+	static inline void clear_bit_##__reg_bit(uintptr_t addr)		\
+	{								\
+		Z_REG_CLEAR_BIT(addr + __reg_off, __bit);		\
+	}
+
+#define DEFINE_TEST_BIT_OP(__reg_bit, __reg_off, __bit)			\
+	static inline int test_bit_##__reg_bit(uintptr_t addr)		\
+	{								\
+		return Z_REG_TEST_BIT(addr + __reg_off, __bit);		\
+	}
+
+#ifdef __cplusplus
+}
+#endif
+
+#if defined(CONFIG_I2C_DW_DEV_RECOVERY)
+int i2c_check_and_recover(const struct device *dev);
+void i2c_platform_spec_init(const struct device *dev);
+#else
+static inline int i2c_check_and_recover(const struct device *dev)
+{
+	return 1;
+}
+static void i2c_platform_spec_init(const struct device *dev) {}
+#endif
+
+#endif /* ZEPHYR_DRIVERS_I2C_I2C_DW_H_ */
diff --git a/drivers/i2c/gaudi3/i2c_dw.h b/drivers/i2c/gaudi3/i2c_dw.h
new file mode 100644
index 0000000..8eaff72
--- /dev/null
+++ b/drivers/i2c/gaudi3/i2c_dw.h
@@ -0,0 +1,225 @@
+/* dw_i2c.h - header for Design Ware I2C operations */
+
+/*
+ * Copyright (c) 2015 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_DRIVERS_I2C_I2C_DW_H_
+#define ZEPHYR_DRIVERS_I2C_I2C_DW_H_
+
+#include <zephyr/drivers/i2c.h>
+#include <stdbool.h>
+
+#define DT_DRV_COMPAT snps_designware_i2c
+
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(pcie)
+BUILD_ASSERT(IS_ENABLED(CONFIG_PCIE), "DW I2C in DT needs CONFIG_PCIE");
+#include <zephyr/drivers/pcie/pcie.h>
+#endif
+
+#if defined(CONFIG_RESET)
+#include <zephyr/drivers/reset.h>
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define I2C_DW_MAGIC_KEY			0x44570140
+
+
+typedef void (*i2c_isr_cb_t)(const struct device *port);
+
+
+#define IC_ACTIVITY                     (1 << 0)
+#define IC_ENABLE_BIT                   (1 << 0)
+
+
+/* dev->state values from IC_DATA_CMD Data transfer mode settings (bit 8) */
+#define I2C_DW_STATE_READY                 (0)
+#define I2C_DW_CMD_SEND                    (1 << 0)
+#define I2C_DW_CMD_RECV                    (1 << 1)
+#define I2C_DW_CMD_ERROR                   (1 << 2)
+#define I2C_DW_BUSY                        (1 << 3)
+
+
+#define DW_ENABLE_TX_INT_I2C_MASTER		(DW_INTR_STAT_TX_OVER |  \
+						 DW_INTR_STAT_TX_EMPTY | \
+						 DW_INTR_STAT_TX_ABRT |  \
+						 DW_INTR_STAT_STOP_DET)
+#define DW_ENABLE_RX_INT_I2C_MASTER		(DW_INTR_STAT_RX_UNDER | \
+						 DW_INTR_STAT_RX_OVER |  \
+						 DW_INTR_STAT_RX_FULL | \
+						 DW_INTR_STAT_STOP_DET)
+
+#define DW_ENABLE_TX_INT_I2C_SLAVE		(DW_INTR_STAT_RD_REQ | \
+						 DW_INTR_STAT_TX_ABRT | \
+						 DW_INTR_STAT_STOP_DET)
+#define DW_ENABLE_RX_INT_I2C_SLAVE		(DW_INTR_STAT_RX_UNDER | \
+						 DW_INTR_STAT_RX_FULL | \
+						 DW_INTR_STAT_RX_DONE | \
+						 DW_INTR_STAT_STOP_DET)
+
+#define DW_DISABLE_ALL_I2C_INT		0x00000000
+
+
+/* IC_CON Low count and high count default values */
+#define I2C_STANDARD_FREQ_KHZ	100
+#define I2C_FAST_FREQ_KHZ	400
+#define I2C_HIGH_FREQ_KHZ	1000
+#define I2C_INPUT_FREQ_MHZ	50
+#define I2C_STD_HCNT		((((I2C_INPUT_FREQ_MHZ * 1000) / I2C_STANDARD_FREQ_KHZ) / 2) - 17)
+#define I2C_STD_LCNT		((((I2C_INPUT_FREQ_MHZ * 1000) / I2C_STANDARD_FREQ_KHZ) / 2) - 1)
+#define I2C_FS_HCNT		((((I2C_INPUT_FREQ_MHZ * 1000) / I2C_FAST_FREQ_KHZ) / 2) - 17)
+#define I2C_FS_LCNT		((((I2C_INPUT_FREQ_MHZ * 1000) / I2C_FAST_FREQ_KHZ) / 2) - 1)
+#define I2C_HS_HCNT		((((I2C_INPUT_FREQ_MHZ * 1000) / I2C_HIGH_FREQ_KHZ) / 2) - 17)
+#define I2C_HS_LCNT		((((I2C_INPUT_FREQ_MHZ * 1000) / I2C_HIGH_FREQ_KHZ) / 2) - 1)
+#define I2C_CUSTOM_HCNT		((CONFIG_I2C_DW_CLOCK_SPEED * 6) / 8)
+#define I2C_CUSTOM_LCNT		((CONFIG_I2C_DW_CLOCK_SPEED * 7) / 8)
+/*
+ * DesignWare speed values don't directly translate from the Zephyr speed
+ * selections in include/i2c.h so here we do a little translation
+ */
+#define I2C_DW_SPEED_STANDARD		0x1
+#define I2C_DW_SPEED_FAST		0x2
+#define I2C_DW_SPEED_FAST_PLUS		0x2
+#define I2C_DW_SPEED_HIGH		0x3
+
+/*
+ * SDA Holdtime conversion from nanosecond to clock ticks.
+ * (0.5*10^6) added before division with 10^6 to avoid
+ * losing fraction.
+ */
+#define i2c_dw_ns_to_clks(x) (((x * CONFIG_I2C_DW_CLOCK_SPEED * 1000) + 500000)\
+								/ 1000000)
+
+
+/*
+ * These values have been randomly selected.  It would be good to test different
+ * watermark levels for performance capabilities
+ */
+#define I2C_DW_TX_WATERMARK		2
+#define I2C_DW_RX_WATERMARK		7
+
+
+struct i2c_dw_rom_config {
+	DEVICE_MMIO_ROM;
+	i2c_isr_cb_t	config_func;
+	uint32_t		bitrate;
+
+#if defined(CONFIG_PINCTRL)
+	const struct pinctrl_dev_config *pcfg;
+#endif
+#if defined(CONFIG_RESET)
+	const struct reset_dt_spec reset;
+#endif
+
+	uint32_t		sda_hold_time;
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(pcie)
+	struct pcie_dev *pcie;
+#endif /* I2C_DW_PCIE_ENABLED */
+
+#ifdef CONFIG_I2C_DW_LPSS_DMA
+	const struct device *dma_dev;
+#endif
+};
+
+struct i2c_dw_dev_config {
+	DEVICE_MMIO_RAM;
+	struct k_sem		device_sync_sem;
+	struct k_mutex		bus_mutex;
+	uint32_t app_config;
+
+	uint8_t			*xfr_buf;
+	uint32_t		xfr_len;
+	uint32_t		rx_pending;
+	bool			user_abort_pending;
+
+	uint16_t		hcnt;
+	uint16_t		lcnt;
+
+	volatile uint8_t	state;  /* last direction of transfer */
+	uint32_t		request_bytes;
+	uint8_t			xfr_flags;
+	bool			support_hs_mode;
+#ifdef CONFIG_I2C_DW_LPSS_DMA
+	uintptr_t phy_addr;
+	uintptr_t base_addr;
+	/* For dma transfer */
+	bool xfr_status;
+#endif
+
+	uint32_t		sda_hold_time_val;
+#ifdef CONFIG_I2C_TARGET
+	struct i2c_target_config *slave_cfg;
+	bool slave_attached;
+#endif
+
+	struct {
+		uint32_t	enable;
+		uint32_t	con;
+		uint32_t	intr_stat;
+		uint32_t	raw_intr_stat;
+		uint32_t	status;
+		uint32_t	tx_abrt_src;
+		uint32_t	rxtl;
+		uint32_t	txtl;
+		uint32_t	rxflr;
+		uint32_t	txflr;
+		uint32_t	comp_param_1;
+	} regs;
+	const char *failure_reason;
+};
+
+#define Z_REG_READ(__sz) sys_read##__sz
+#define Z_REG_WRITE(__sz) sys_write##__sz
+#define Z_REG_SET_BIT sys_set_bit
+#define Z_REG_CLEAR_BIT sys_clear_bit
+#define Z_REG_TEST_BIT sys_test_bit
+
+#define DEFINE_MM_REG_READ(__reg, __off, __sz)				\
+	static inline uint32_t read_##__reg(uintptr_t addr)			\
+	{								\
+		return Z_REG_READ(__sz)(addr + __off);			\
+	}
+#define DEFINE_MM_REG_WRITE(__reg, __off, __sz)				\
+	static inline void write_##__reg(uint32_t data, uintptr_t addr)	\
+	{								\
+		Z_REG_WRITE(__sz)(data, addr + __off);			\
+	}
+
+#define DEFINE_SET_BIT_OP(__reg_bit, __reg_off, __bit)			\
+	static inline void set_bit_##__reg_bit(uintptr_t addr)		\
+	{								\
+		Z_REG_SET_BIT(addr + __reg_off, __bit);			\
+	}
+
+#define DEFINE_CLEAR_BIT_OP(__reg_bit, __reg_off, __bit)		\
+	static inline void clear_bit_##__reg_bit(uintptr_t addr)		\
+	{								\
+		Z_REG_CLEAR_BIT(addr + __reg_off, __bit);		\
+	}
+
+#define DEFINE_TEST_BIT_OP(__reg_bit, __reg_off, __bit)			\
+	static inline int test_bit_##__reg_bit(uintptr_t addr)		\
+	{								\
+		return Z_REG_TEST_BIT(addr + __reg_off, __bit);		\
+	}
+
+#ifdef __cplusplus
+}
+#endif
+
+#if defined(CONFIG_I2C_DW_DEV_RECOVERY)
+int i2c_check_and_recover(const struct device *dev);
+void i2c_platform_spec_init(const struct device *dev);
+#else
+static inline int i2c_check_and_recover(const struct device *dev)
+{
+	return 1;
+}
+static void i2c_platform_spec_init(const struct device *dev) {}
+#endif
+
+#endif /* ZEPHYR_DRIVERS_I2C_I2C_DW_H_ */
diff --git a/drivers/i2c/i2c_dw.c b/drivers/i2c/i2c_dw.c
index e94c917..8cb9d18 100644
--- a/drivers/i2c/i2c_dw.c
+++ b/drivers/i2c/i2c_dw.c
@@ -979,7 +979,22 @@
 		dw->hcnt = value;
 		break;
 	case I2C_SPEED_FAST:
-		__fallthrough;
+		if (I2C_CUSTOM_LCNT <= (read_hs_spklen(reg_base) + 7)) {
+			value = read_hs_spklen(reg_base) + 8;
+		} else {
+			value = I2C_CUSTOM_LCNT;
+		}
+
+		dw->lcnt = value;
+
+		if (I2C_CUSTOM_HCNT <= (read_hs_spklen(reg_base) + 5)) {
+			value = read_hs_spklen(reg_base) + 6;
+		} else {
+			value = I2C_CUSTOM_HCNT;
+		}
+
+		dw->hcnt = value;
+		break;
 	case I2C_SPEED_FAST_PLUS:
 		/*
 		 * Following the directions on DW spec page 59, IC_FS_SCL_LCNT
@@ -1095,7 +1110,7 @@
 	dw->app_config = i2c_map_dt_bitrate(rom->bitrate);
 
 	if (i2c_dw_runtime_configure(dev, dw->app_config) != 0) {
-		LOG_DBG("I2C: Cannot set default configuration");
+		LOG_ERR("I2C: Cannot set default configuration");
 		return -EIO;
 	}
 
diff --git a/drivers/i2c/i2c_dw.h b/drivers/i2c/i2c_dw.h
deleted file mode 100644
index 0694f8b..0000000
--- a/drivers/i2c/i2c_dw.h
+++ /dev/null
@@ -1,221 +0,0 @@
-/* dw_i2c.h - header for Design Ware I2C operations */
-
-/*
- * Copyright (c) 2015 Intel Corporation
- *
- * SPDX-License-Identifier: Apache-2.0
- */
-#ifndef ZEPHYR_DRIVERS_I2C_I2C_DW_H_
-#define ZEPHYR_DRIVERS_I2C_I2C_DW_H_
-
-#include <zephyr/drivers/i2c.h>
-#include <stdbool.h>
-
-#define DT_DRV_COMPAT snps_designware_i2c
-
-#if DT_ANY_INST_ON_BUS_STATUS_OKAY(pcie)
-BUILD_ASSERT(IS_ENABLED(CONFIG_PCIE), "DW I2C in DT needs CONFIG_PCIE");
-#include <zephyr/drivers/pcie/pcie.h>
-#endif
-
-#if defined(CONFIG_RESET)
-#include <zephyr/drivers/reset.h>
-#endif
-
-#ifdef __cplusplus
-extern "C" {
-#endif
-
-#define I2C_DW_MAGIC_KEY			0x44570140
-
-
-typedef void (*i2c_isr_cb_t)(const struct device *port);
-
-
-#define IC_ACTIVITY                     (1 << 0)
-#define IC_ENABLE_BIT                   (1 << 0)
-
-
-/* dev->state values from IC_DATA_CMD Data transfer mode settings (bit 8) */
-#define I2C_DW_STATE_READY                 (0)
-#define I2C_DW_CMD_SEND                    (1 << 0)
-#define I2C_DW_CMD_RECV                    (1 << 1)
-#define I2C_DW_CMD_ERROR                   (1 << 2)
-#define I2C_DW_BUSY                        (1 << 3)
-
-
-#define DW_ENABLE_TX_INT_I2C_MASTER		(DW_INTR_STAT_TX_OVER |  \
-						 DW_INTR_STAT_TX_EMPTY | \
-						 DW_INTR_STAT_TX_ABRT |  \
-						 DW_INTR_STAT_STOP_DET)
-#define DW_ENABLE_RX_INT_I2C_MASTER		(DW_INTR_STAT_RX_UNDER | \
-						 DW_INTR_STAT_RX_OVER |  \
-						 DW_INTR_STAT_RX_FULL | \
-						 DW_INTR_STAT_STOP_DET)
-
-#define DW_ENABLE_TX_INT_I2C_SLAVE		(DW_INTR_STAT_RD_REQ | \
-						 DW_INTR_STAT_TX_ABRT | \
-						 DW_INTR_STAT_STOP_DET)
-#define DW_ENABLE_RX_INT_I2C_SLAVE		(DW_INTR_STAT_RX_UNDER | \
-						 DW_INTR_STAT_RX_FULL | \
-						 DW_INTR_STAT_RX_DONE | \
-						 DW_INTR_STAT_STOP_DET)
-
-#define DW_DISABLE_ALL_I2C_INT		0x00000000
-
-
-/* IC_CON Low count and high count default values */
-/* TODO verify values for high and fast speed */
-#define I2C_STD_HCNT			(CONFIG_I2C_DW_CLOCK_SPEED * 4)
-#define I2C_STD_LCNT			(CONFIG_I2C_DW_CLOCK_SPEED * 5)
-#define I2C_FS_HCNT			((CONFIG_I2C_DW_CLOCK_SPEED * 6) / 8)
-#define I2C_FS_LCNT			((CONFIG_I2C_DW_CLOCK_SPEED * 7) / 8)
-#define I2C_HS_HCNT			((CONFIG_I2C_DW_CLOCK_SPEED * 6) / 8)
-#define I2C_HS_LCNT			((CONFIG_I2C_DW_CLOCK_SPEED * 7) / 8)
-
-/*
- * DesignWare speed values don't directly translate from the Zephyr speed
- * selections in include/i2c.h so here we do a little translation
- */
-#define I2C_DW_SPEED_STANDARD		0x1
-#define I2C_DW_SPEED_FAST		0x2
-#define I2C_DW_SPEED_FAST_PLUS		0x2
-#define I2C_DW_SPEED_HIGH		0x3
-
-/*
- * SDA Holdtime conversion from nanosecond to clock ticks.
- * (0.5*10^6) added before division with 10^6 to avoid
- * losing fraction.
- */
-#define i2c_dw_ns_to_clks(x) (((x * CONFIG_I2C_DW_CLOCK_SPEED * 1000) + 500000)\
-								/ 1000000)
-
-
-/*
- * These values have been randomly selected.  It would be good to test different
- * watermark levels for performance capabilities
- */
-#define I2C_DW_TX_WATERMARK		2
-#define I2C_DW_RX_WATERMARK		7
-
-
-struct i2c_dw_rom_config {
-	DEVICE_MMIO_ROM;
-	i2c_isr_cb_t	config_func;
-	uint32_t		bitrate;
-
-#if defined(CONFIG_PINCTRL)
-	const struct pinctrl_dev_config *pcfg;
-#endif
-#if defined(CONFIG_RESET)
-	const struct reset_dt_spec reset;
-#endif
-
-	uint32_t		sda_hold_time;
-#if DT_ANY_INST_ON_BUS_STATUS_OKAY(pcie)
-	struct pcie_dev *pcie;
-#endif /* I2C_DW_PCIE_ENABLED */
-
-#ifdef CONFIG_I2C_DW_LPSS_DMA
-	const struct device *dma_dev;
-#endif
-};
-
-struct i2c_dw_dev_config {
-	DEVICE_MMIO_RAM;
-	struct k_sem		device_sync_sem;
-	struct k_mutex		bus_mutex;
-	uint32_t app_config;
-
-	uint8_t			*xfr_buf;
-	uint32_t		xfr_len;
-	uint32_t		rx_pending;
-	bool			user_abort_pending;
-
-	uint16_t		hcnt;
-	uint16_t		lcnt;
-
-	volatile uint8_t	state;  /* last direction of transfer */
-	uint32_t		request_bytes;
-	uint8_t			xfr_flags;
-	bool			support_hs_mode;
-#ifdef CONFIG_I2C_DW_LPSS_DMA
-	uintptr_t phy_addr;
-	uintptr_t base_addr;
-	/* For dma transfer */
-	bool xfr_status;
-#endif
-
-	uint32_t		sda_hold_time_val;
-#ifdef CONFIG_I2C_TARGET
-	struct i2c_target_config *slave_cfg;
-	bool slave_attached;
-#endif
-
-	struct {
-		uint32_t	enable;
-		uint32_t	con;
-		uint32_t	intr_stat;
-		uint32_t	raw_intr_stat;
-		uint32_t	status;
-		uint32_t	tx_abrt_src;
-		uint32_t	rxtl;
-		uint32_t	txtl;
-		uint32_t	rxflr;
-		uint32_t	txflr;
-		uint32_t	comp_param_1;
-	} regs;
-	const char *failure_reason;
-};
-
-#define Z_REG_READ(__sz) sys_read##__sz
-#define Z_REG_WRITE(__sz) sys_write##__sz
-#define Z_REG_SET_BIT sys_set_bit
-#define Z_REG_CLEAR_BIT sys_clear_bit
-#define Z_REG_TEST_BIT sys_test_bit
-
-#define DEFINE_MM_REG_READ(__reg, __off, __sz)				\
-	static inline uint32_t read_##__reg(uintptr_t addr)			\
-	{								\
-		return Z_REG_READ(__sz)(addr + __off);			\
-	}
-#define DEFINE_MM_REG_WRITE(__reg, __off, __sz)				\
-	static inline void write_##__reg(uint32_t data, uintptr_t addr)	\
-	{								\
-		Z_REG_WRITE(__sz)(data, addr + __off);			\
-	}
-
-#define DEFINE_SET_BIT_OP(__reg_bit, __reg_off, __bit)			\
-	static inline void set_bit_##__reg_bit(uintptr_t addr)		\
-	{								\
-		Z_REG_SET_BIT(addr + __reg_off, __bit);			\
-	}
-
-#define DEFINE_CLEAR_BIT_OP(__reg_bit, __reg_off, __bit)		\
-	static inline void clear_bit_##__reg_bit(uintptr_t addr)		\
-	{								\
-		Z_REG_CLEAR_BIT(addr + __reg_off, __bit);		\
-	}
-
-#define DEFINE_TEST_BIT_OP(__reg_bit, __reg_off, __bit)			\
-	static inline int test_bit_##__reg_bit(uintptr_t addr)		\
-	{								\
-		return Z_REG_TEST_BIT(addr + __reg_off, __bit);		\
-	}
-
-#ifdef __cplusplus
-}
-#endif
-
-#if defined(CONFIG_I2C_DW_DEV_RECOVERY)
-int i2c_check_and_recover(const struct device *dev);
-void i2c_platform_spec_init(const struct device *dev);
-#else
-static inline int i2c_check_and_recover(const struct device *dev)
-{
-	return 1;
-}
-static void i2c_platform_spec_init(const struct device *dev) {}
-#endif
-
-#endif /* ZEPHYR_DRIVERS_I2C_I2C_DW_H_ */
diff --git a/dts/arc/hl_gaudi3.dtsi b/dts/arc/hl_gaudi3.dtsi
index 20492b9..d10af7b 100644
--- a/dts/arc/hl_gaudi3.dtsi
+++ b/dts/arc/hl_gaudi3.dtsi
@@ -512,7 +512,7 @@
 			#size-cells = <0>;
 			reg = <0xfc481000 0x100>;
 			interrupts = <321 2>;
-			clock-frequency = <I2C_BITRATE_FAST_PLUS>;
+			clock-frequency = <I2C_BITRATE_FAST>;
 			pinctrl-0  = <&i2c1_default>;
 			pinctrl-names = "default";
 			node_label = "I2C_1";
@@ -615,7 +615,7 @@
 			interrupts = <322 1>;
 			pinctrl-0 = <&i2c2_default>;
 			pinctrl-names = "default";
-			clock-frequency = <I2C_BITRATE_FAST_PLUS>;
+			clock-frequency = <I2C_BITRATE_STANDARD>;
 			node_label = "I2C_2";
 			slave_soc0: slave-soc@67 {
 				status = "disabled";
